draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -669,65 +883,101 @@ Algorithms in this document make use of utility functions described below.
-from a 256-bit output, the values from 2^256 - 38 to 2^256 - 1 have more
-probability to be chosen. In this example, the resulting bias is negligible,
-but for others this bias could be significant. This is a standard problem in
-generating uniformly distributed integers from a bitstring. In order to smooth
-out this bias, the hash2base algorithm should use a hash function that produces
-as many bits as possible before reducing modulo p. This approach is preferable
-to an iterated procedure, such as rejection sampling, since this can be hard to reliably implement in constant time.
-
-Performance: hash2base requires hashing the entire input x. In some algorithms
-or ciphersuite combinations, hash2base is called multiple times. For large
-inputs, implementers can therefore consider hashing x before calling hash2base, i.e., setting x=H(x).
+The hash2base(msg) function maps a string msg of any length into an element of a
+field F. This function is parametrized by the field F ({{bg-curves}}) and by H,
+a cryptographic hash function that outputs b bits.
+
+At a high level, for q = p^m the order of F, hash2base(msg) outputs a vector of m elements,
+each an integer modulo p (see {{bg-curves}} for a brief discussion of
+representing elements in extension fields).
+
+## Security and performance considerations
+
+Hash2base should map its input to a uniformly random element of F.
+In practice, however, the output may be biased, meaning that some field
+elements are more likely to occur than others. This is true even when H
+outputs a uniformly random string (as is generally assumed). For example,
+if H=SHA256 and F is a field of characteristic p = 2^255 - 19, then the
+result of reducing H(msg) (a 256-bit integer) modulo p is slightly more likely
+to be a value in \[0, 38\] than a value in \[39, 2^255 - 19). In this example
+the bias is negligible, but in general the bias can be significant.
+
+To control bias, the input msg should be hashed to an integer comprising more than log2(p) bits.
+In particular, reducing an integer of ceil(log2(p)) + k bits modulo p gives bias
+at most 2^-k, which is a safe choice for a cryptosystem with k-bit security.
+Thus, if H outputs b bits, then H should be evaluated W = ceil((ceil(log2(p)) + k) / b)
+times to produce a W * b bit integer, which is then reduced modulo p.
+
+Note that implementors SHOULD NOT use an iterated procedure, i.e., rejection
+sampling. The reason is that these procedures are difficult to implement in constant time,
+and later well-meaning "optimizations" may silently render an implementation
+non-constant-time.
+
+The performance of hash2base may be limited by the length of the input m.
+To prevent this, hash2base first computes m' = H(msg) and then derives the
+required bits from m'. This entails one extra invocation of H, a
+negligible overhead.
 
 ## Implementation
 
 The following procedure implements hash2base.
 
 ~~~
-hash2base(x)
+hash2base(msg)
```

Comment:
s/`msg`/`alpha` to be consistent with use of `hash2base` in the rest of the document?

Reply:
Sure. Is it OK if I roll this into the changes to hash2base that we discussed via email (referenced in a comment above)?

Reply:
Also: do you think a global msg -> alpha replacement, or just local to this definition?

(Also, I suppose it's worth mentioning the other obvious option: we could do alpha -> msg instead...)

Reply:
Hmm, I think just doing it locally is probably fine for now. Though I'm curious to hear what @armfazh and @samscott89 think.

Reply:
How about we take this into a separate issue / discussion / PR? It would be good to settle on conventions throughout the document (variable names, spacing, etc.), and that's more or less orthogonal to settling on the content.

Reply:
Created Issue #110.